File: bi/bi-operations/src/main/java/com/yazino/bi/operations/controller/LoginController.java
package com.yazino.bi.operations.controller;

import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.servlet.ModelAndView;

import javax.servlet.http.HttpSession;

@Controller
@RequestMapping("/login")
public class LoginController {

    @RequestMapping
    public String login() {
        return "login";
    }

    @RequestMapping("/error")
    public ModelAndView loginError(final HttpSession session) {
        return new ModelAndView("login")
                .addObject("error", lastSpringSecurityError(session));
    }

    private String lastSpringSecurityError(final HttpSession session) {
        final Exception lastException = (Exception) session.getAttribute("SPRING_SECURITY_LAST_EXCEPTION");
        if (lastException != null) {
            return lastException.getMessage();
        }
        return null;
    }

}

File: platform/control-centre/src/main/scala/strata/server/lobby/controlcentre/controller/LoginController.scala
package strata.server.lobby.controlcentre.controller

import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.servlet.ModelAndView
import javax.servlet.http.HttpSession

@Controller
class LoginController {

    @RequestMapping(Array("login"))
    def login(): String = "login"

    @RequestMapping(Array("login/error"))
    def home(session: HttpSession): ModelAndView = {
        new ModelAndView("login")
                .addObject("error", lastSpringSecurityError(session))
    }

    private def lastSpringSecurityError(session: HttpSession): String = {
        val lastException: Exception = session.getAttribute("SPRING_SECURITY_LAST_EXCEPTION").asInstanceOf[Exception]
        if (lastException != null) {
            lastException.getMessage
        } else {
            null
        }
    }
}
